### PR TITLE
[MLIR] Vector: turn the ExtractStridedSlice rewrite pattern from #111541 into a canonicalization

### PR DESCRIPTION
**Pull request by bjacob**
This is a reasonable canonicalization because `extract` is more constrained than `extract_strided_slices`, so there is no loss of semantics here, just lifting an op to a special-case higher/constrained op. And the additional `shape_cast` is merely adding leading unit dims to match the original result type.

Context: discussion on #111541. I wasn't sure how this would turn out, but in the process of writing this PR, I discovered at least 2 bugs in the pattern introduced in #111541, which shows the value of shared canonicalization patterns which are exercised on a high number of testcases.

### CODE REVIEW

File: mlir/lib/Dialect/Vector/IR/VectorOps.cpp [hunk 1]
```diff
@@ -3772,6 +3772,92 @@ class StridedSliceSplat final : public OpRewritePattern<ExtractStridedSliceOp> {
...
+    if (numOffsets == sourceType.getRank() &&
+        static_cast<int>(sizes.size()) == sourceType.getRank()) {
+      return failure();
+    }
+
+    // The outer dimensions must have unit size.
+    for (int i = 0; i < numOffsets; ++i) {
+      if (sizes[i] != 1) {
+        return failure();
+      }
```
Comment (joker-eph) [inline mlir/lib/Dialect/Vector/IR/VectorOps.cpp:3834]: Nit: extra braces on this if (and the 3 others above)

File: mlir/lib/Dialect/Vector/IR/VectorOps.cpp [hunk 2]
```diff
@@ -3772,6 +3772,92 @@ class StridedSliceSplat final : public OpRewritePattern<ExtractStridedSliceOp> {
...
+/// Example:
+///     Before:
+///         %1 = vector.extract_strided_slice %arg0 {offsets = [0, 0, 0, 0, 0],
+///         sizes = [1, 1, 1, 1, 8], strides = [1, 1, 1, 1, 1]} :
+///         vector<8x1x1x2x8xi8> to vector<1x1x1x1x8xi8>
+///     After:
+///         %0 = vector.extract %arg0[0, 0, 0, 0] : vector<8xi8> from
+///         vector<8x1x1x2x8xi8> %1 = vector.shape_cast %0 : vector<8xi8> to
+///         vector<1x1x1x1x8xi8>
+///
```
Comment (joker-eph) [inline mlir/lib/Dialect/Vector/IR/VectorOps.cpp:3787]: Maybe document the casts insertion as well?

**Review by joker-eph — APPROVED** [review]
Looks good to me, but just please update the description to justify why it is a good canonicalization and state that we don't see any loss of semantics in the transformation here.

Thanks!

File: mlir/lib/Dialect/Vector/IR/VectorOps.cpp (hunk 2 above)
Comment (bjacob) [reply mlir/lib/Dialect/Vector/IR/VectorOps.cpp:3790]: That was clang-format eating the newline and making it look like a single op.

**Comment by bjacob** [comment]
The 2nd commit pushed here addresses review comments AND fixes a bug (thankfully just by deleting a few lines).  I was over-zealously trying to prevent leading unit dims in the `extract` result type and in the process I was causing `numOffsets` to exceed `offsets.size()` which would cause an out-of-bounds read.

**Review by banach-space — APPROVED** [review]
Makes sense, thanks!

**Pull request merged**